src/mon/OSDMonitor.cc

```diff
@@ -1127,6 +1127,48 @@ void OSDMonitor::encode_pending(MonitorDBStore::TransactionRef t)
 			<< "required " << ceph_release_name(mv);
       pending_inc.new_require_min_compat_client = mv;
     }
+
+    // upgrade to mimic?
+    if (osdmap.require_osd_release < CEPH_RELEASE_MIMIC &&
+	tmp.require_osd_release >= CEPH_RELEASE_MIMIC) {
+      dout(10) << __func__ << " first mimic epoch" << dendl;
+      // record this epoch as the deletion for all legacy removed_snaps
+      for (auto& p : tmp.get_pools()) {
+	// update every pool
+	if (pending_inc.new_pools.count(p.first) == 0) {
+	  pending_inc.new_pools[p.first] = p.second;
+	}
+	auto& pi = pending_inc.new_pools[p.first];
+	if (pi.snap_seq > 0 &&
+	    (pi.flags & (pg_pool_t::FLAG_SELFMANAGED_SNAPS |
+			 pg_pool_t::FLAG_POOL_SNAPS)) == 0) {
+	  if (!pi.removed_snaps.empty()) {
+	    pi.flags |= pg_pool_t::FLAG_SELFMANAGED_SNAPS;
+	  } else {
+	    pi.flags |= pg_pool_t::FLAG_POOL_SNAPS;
+	  }
+	}
+
+	if (!p.second.removed_snaps.empty()) {
+	  // all snaps removed this epoch
+	  dout(10) << __func__ << " converting pool " << p.first
+		   << " with " << p.second.removed_snaps.size()
+		   << " legacy removed_snaps" << dendl;
+	  string k = make_snap_epoch_key(p.first, pending_inc.epoch);
+	  bufferlist v;
+	  ::encode(p.second.removed_snaps, v);
+	  t->put(OSD_SNAP_PREFIX, k, v);
+	  for (auto q = p.second.removed_snaps.begin();
+	       q != p.second.removed_snaps.end();
+	       ++q) {
+	    bufferlist v;
+	    string k = make_snap_key_value(p.first, q.get_start(),
+					   q.get_len(), pending_inc.epoch, &v);
```

Comment:
The same information is encoded in the pg_pool_t in the OSDMap *every* time the pool metadata changes (i.e. a snap is created or removed).  This is somewhat more expensive in that it is doing many keys instead of one big value.  For earlier versions of this patch I had some test code that generated huuuge sparse removed_sets and tested and it worked fine (based on an osdmap sample I got from that guy submitting the interval_set patches).  This is more efficient than that code was bc it is interval-based instead of vector-based.  I think it'll be okay.  It's also a one-time thing.. so even if they get a mon election blip it will clear up.

Reply:
👍